### PR TITLE
Adds the mystery box

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
A compliment to my recent root beer PR, since I've been on a retro CoD Zombies kick lately, this PR adds a new admin structure that randomly rolls an item and presents it to the player like the mystery box from... well, CoD Zombies.

[Here's the new public domain version in action.](https://streamable.com/01btp8) The music that plays while selecting is The Ragtime Drummer by James Lent, sourced from [freemusicarchive](https://freemusicarchive.org/music/Antique_Phonograph_Music_Program_Various_Artists/Antique_Phonograph_Music_Program_11032009/James_Lent_-_The_Ragtime_Drummer).

There's a base version that can spawn any item that the christmas tree/santa's bag can spawn, and then there's a subtype that draws from the Summon Guns list only. If you want, you can use the base type to generate a list of any subtypes of a given path, or you can edit the valid_types list to only consider specific types of atoms.

## Why It's Good For The Game
Fun in the thunderdome, giving players a fun way to spawn in weapons.

## Changelog

:cl: Ryll/Shaps
add: Added in a new admin spawn, the mystery box! Dealing randomized (maybe) death to thunderdomes near you! Maybe!
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
@Ryll-Ryll stop playing so much kino der toten i think it's starting to affect your pr ideas

**Comment by Jackraxxus** [comment]
If this is meant for thunderdome gaming could you maybe blacklist the staff/wand of safety?
It teleports people hit to the station.

**Comment by Ryll-Ryll** [comment]
> @Ryll-Ryll stop playing so much kino der toten i think it's starting to affect your pr ideas

pack a punch next

> If this is meant for thunderdome gaming could you maybe blacklist the staff/wand of safety? It teleports people hit to the station.

Yeah, I'll include some presets and easy tools to change them shortly

**Comment by AndroidMK2** [comment]
Start playing barotrauma so we get our sexy underwater map

**Comment by StoneMason** [comment]
it should take 950 credits to use

**Comment by optimumtact** [comment]
@Ryll-Ryll free to download doesn't necessarily translate into free to use, do you know what actual license it's under?

**Comment by Ryll-Ryll** [comment]
Alright clearly I was misled by youtube comments about Slapstick Dick's copyright status, I've replaced it with a new public domain tune called The Ragtime Drummer by James Lent. Also added some new functionality, so it can currently act like a somewhat flashier yet slower christmas tree, as well as a gun dispenser

File: code/game/objects/structures/mystery_box.dm [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+/obj/structure/mystery_box/Initialize(mapload)
+	. = ..()
+	generate_valid_types()
+
+/obj/structure/mystery_box/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+	switch(box_state)
+		if(MYSTERY_BOX_COOLING_DOWN)
+
```
Comment (tralezab) [inline code/game/objects/structures/mystery_box.dm:105]: remove empty switch cases, i know why they're there but they don't really help reading that much

File: code/game/objects/structures/mystery_box.dm [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+		if(MYSTERY_BOX_STANDBY)
+			activate(user)
+
+		if(MYSTERY_BOX_CHOOSING)
+
```
Comment (tralezab) [inline code/game/objects/structures/mystery_box.dm:110]: see above

File: code/game/objects/structures/mystery_box.dm [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+/obj/structure/mystery_box/update_icon_state()
+	icon_state = "[initial(icon_state)][box_state > MYSTERY_BOX_STANDBY ? "open" : ""]"
+	return ..()
+
+/// This proc is used to define what item types valid_types is filled with
+/obj/structure/mystery_box/proc/generate_valid_types()
+	valid_types = list()
+
+	for(var/iter_path in typesof(selectable_base_type))
+		if(ispath(iter_path) && ispath(iter_path, /obj/item))
```
Comment (tralezab) [inline code/game/objects/structures/mystery_box.dm:124]: early continue, why not?

File: code/game/objects/structures/mystery_box.dm [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+/obj/mystery_box_item/proc/start_animation(atom/parent)
+	parent_box = parent
+
+	var/matrix/starting = matrix()
+	animate(src, pixel_y = 6, transform = starting, time = MBOX_DURATION_CHOOSING, easing = QUAD_EASING | EASE_OUT)
+	loop_icon_changes()
+
+/// Keep changing the icon and selected path
+/obj/mystery_box_item/proc/loop_icon_changes()
+	set waitfor = FALSE
```
Comment (tralezab) [inline code/game/objects/structures/mystery_box.dm:243]: INVOKE_ASYNC the proc instead

**Review by tralezab — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/mystery_box.dm [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+		selected_path = pick(parent_box.valid_types)
+		selected_item = selected_path
+		icon = initial(selected_item.icon)
+		icon_state = initial(selected_item.icon_state)
+		sleep(change_delay)
+
+	add_filter("ready_outline", 2, list("type" = "outline", "color" = "#FBFF23", "size" = 0.2))
+	name = initial(selected_item.name)
+	parent_box.present_weapon()
+	claimable = TRUE
```
Comment (LemonInTheDark) [inline code/game/objects/structures/mystery_box.dm:263]: I think these could be a series of animates and a timer, would make life a bit easier

Comment (LemonInTheDark) [reply code/game/objects/structures/mystery_box.dm]: Wow it just doesn't work. that's depressing

Comment (Ryll-Ryll) [reply code/game/objects/structures/mystery_box.dm]: Yeah I was messing with it over the last few days with different configurations, but it appears to just take the last one and run it immediately. I'm not thrilled about using all timers instead but meh

**Comment by cacogen** [comment]
bioshock vibes

**Pull request merged**